js/modules/k6/experimental/fs/cache.go

```diff
@@ -0,0 +1,86 @@
+package fs
+
+import (
+	"fmt"
+	"io"
+	"path/filepath"
+	"sync"
+
+	"go.k6.io/k6/lib/fsext"
+)
+
+// cache is a cache of opened files.
+type cache struct {
+	// openedFiles holds a safe for concurrent use map, holding the content
+	// of the files that were opened by the user.
+	//
+	// Keys are expected to be strings holding the openedFiles' path.
+	// Values are expected to be byte slices holding the content of the opened file.
+	//
+	// That way, we can cache the file's content and avoid opening too many
+	// file descriptor, and re-reading its content every time the file is opened.
+	//
+	// Importantly, this also means that if the
+	// file is modified from outside of k6, the changes will not be reflected in the file's data.
+	openedFiles sync.Map
+}
+
+// open retrieves the content of a given file from the specified filesystem (fromFs) and
+// stores it in the registry's internal `openedFiles` map.
+//
+// The function cleans the provided filename using filepath.Clean before using it.
+//
+// If the file was previously "opened" (and thus cached) by the registry, it
+// returns the cached content. Otherwise, it reads the file from the
+// filesystem, caches its content, and then returns it.
+//
+// The function is designed to minimize redundant file reads by leveraging an internal cache (openedFiles).
+// In case the cached value is not a byte slice (which should never occur in regular use), it
+// panics with a descriptive error.
+//
+// Parameters:
+//   - filename: The name of the file to be retrieved. This should be a relative or absolute path.
+//   - fromFs: The filesystem (from the fsext package) from which the file should be read if not already cached.
+//
+// Returns:
+//   - A byte slice containing the content of the specified file.
+//   - An error if there's any issue opening or reading the file. If the file content is
+//     successfully cached and returned once, subsequent calls will not produce
+//     file-related errors for the same file, as the cached value will be used.
+func (fr *cache) open(filename string, fromFs fsext.Fs) (data []byte, err error) {
+	filename = filepath.Clean(filename)
+
+	if f, ok := fr.openedFiles.Load(filename); ok {
+		data, ok = f.([]byte)
+		if !ok {
+			panic(fmt.Errorf("registry's file %s is not stored as a byte slice", filename))
+		}
+
+		return data, nil
+	}
+
+	// The underlying fsext.Fs.Open method will cache the file content during this
+	// operation. Which will lead to effectively holding the content of the file in memory twice.
+	// However, as per #1079, we plan to eventually reduce our dependency on afero, and
+	// expect this issue to be resolved at that point.
+	// TODO: re-evaluate opening from the FS this once #1079 is resolved.
```

Comment:
I think this comment is missing an explanation about _why_ we did the same thing if it was already possible in another way. And _why_ we ended up doing it twice.

Related to it, generally, I think now that we have implemented it, and in some details, it diverges from the design doc. For this reason, I would like to have somewhere in this code a comment explaining the goals of this module.
We have them here https://github.com/grafana/k6/blob/master/docs/design/019-file-api.md#proposed-solution but I think it would make sense to have them repeated in a comment in the code. Because the design doc is not referenced here, it could be difficult to discover and it could become outdated in the future. I think the code's comments are easier to maintain.

Reply:
I've added a more explicit documentation to the `cache` type, explaining why it exists, referring to how it differs some of the constraints expressed in the design document, and pointing to #1079 as opening future opportunities.

The type comment superseeds the previous one made at the site of the `Open` call which I removed.

I'll mark this as resolved, but feel free to comment further on the newer version introduced by my latest commit 👍🏻